### ISSUE TITLE
Allow to run FluxCD per tenant

### ISSUE THREAD

**Issue by kvaps** [cozystack/cozystack#184]
We want the Flux checkbox right next to etcd, ingress, monitoring

This would allow to move FluxCD to system applications so it will have a packaged Helm-chart to install from

**Comment by kvaps** [comment]
Probably locked by https://github.com/controlplaneio-fluxcd/flux-operator/issues/55

**Comment by kingdonb** [comment]
The design of Flux multi-tenancy is orthogonal from the notion of sharding, or multiple flux instances per cluster.

Sharding was added as a last resort for users who need tens of thousands of HelmReleases per cluster, it is a scaling consideration now more than an accommodation for tenancy.

While it has been considered that you can use it for other purposes like flux-per-tenant (especially where secrets need to be mounted into to the controller eg. AWS IRSA or vault for secrets) those ideas are not really well explored in documentation or tests, there is no support for multi-tenant sharding today.

Yes, the CRDs have some strong guarantees for backwards compatibility, but it doesn't cancel out this coordination issue. 

Users of Flux need to have a `flux` CLI that matches their deployed controllers & CRDs one-to-one. If it doesn't match, you get strange errors now and then - with the extent of the mismatch and the changes between versions, the issues will vary.

So it is a coordination issue. If users on the same cluster are tenants, they should not be expected to coordinate with each other in any way. The idea of a FluxCD per tenant implies that a tenant can choose when to do this upgrade. When one tenant upgrades, they will install CRDs from the next release, all other tenants must either upgrade or suffer the mismatch.

If flux-operator always installs the latest Flux instance, that solves the coordination issue, but that tenants get to decide their own upgrade cadence and are not bound to each other seems like it would be a fundamental purpose of this kind of implementation. And it's unfortunately not as simple as a version skew policy like kubectl, the Flux docs are (or at least should be) quite explicit that the CLI version and distribution installed **need to match**. You can have situations where from one version to the next, new CLI will not work with the CRDs from the prior version because of APIVersion mismatches.

**Comment by gecube** [comment]
we will have the same issue with the tenant cluster versions and compatibility between kubectl utility and tenants clusters...

**Comment by kingdonb** [comment]
Not exactly, kubectl supports a version skew policy of plus-or-minus one:

https://kubernetes.io/releases/version-skew-policy/#kubectl

So you are guaranteed that kubectl 1.29 works with kubelet + api server 1.30

There's no transitional skew guarantee like that for Flux. When you download your new Flux CLI, the next step is bootstrap upgrade and alert your colleagues that it's time to upgrade their CLIs.

The CRD backwards compatibility guarantees of Flux I can't explain in detail right now, but they are based on alpha/beta support guarantees from Kubernetes upstream. Eg. when an API is promoted from alpha to beta, it's considered production ready, and automatic upgrades are supported for one or more versions, until the deprecated API version can be removed.

All Flux APIs are currently beta or later, so those guarantees are in place.

An example where you can have trouble, Flux 2.3 CLI can't work with Flux 2.2 because some deprecated versions are finally removed after years. You will get crazy errors if you try mixing across those versions. On one hand this does not happen very often, (on the other hand it is long in advance predictable that it happens, it does happen, and it will probably happen again.)

**Comment by gecube** [comment]
> Flux 2.3 CLI can't work with Flux 2.2 because some deprecated versions are finally removed after years.

it is totally not understandable, because kubectl works (!)

**Comment by NickVolynkin** [comment]
Meeting notes: not supported in upstream, so we use what we have.
